Replace loading spinner with snippet

eugenekasimov opened the pull request:
### PR Summary: 

This PR is to replace all exciting hard coded loading-overlay spinners with snippets.

### Why are these changes introduced?

Fixes #2954.

### What approach did you take?

I replaced hardcoded `loading-overlay spinners` with snippet in liquid files:

- Feat product
- Main cart items
- Main product
- Buy buttons
- Card product
- Cart drawer
- Product thumbnail

Since the spinners' structures were slightly different from file to file I changed CSS classes accordingly to unify all of them. 

### Other considerations

I noticed that in liquid files below 👇 we had spinners but they never were in action. I removed them with CSS accordingly. 
- Main search 
- Facets
- Main collection product grid

Pls pay attention for this 👆 to  make sure I didn't break anything 😅

### Visual impact on existing themes
<!-- How will this visually affect merchants who upgrade to a new theme version with this change? -->
No visual impact

### Testing steps/scenarios
<!-- List all the testing tasks that applies to your fix to help peers review your work. -->
- [ ] Go over all files mentioned in description and make sure the loading-overlay spinner still appears correctly. 

### Demo links
<!-- Please include a link to a demo store that includes preconfigured sections and settings to allow reviewers to easily test the features you are working on. -->

- [Editor](url)

### Checklist
- [ ] Added PR summary for [release notes](https://themes.shopify.com/themes/dawn/styles/default#ReleaseNotes)
- [ ] Requested review from UX (Only for changes that are affecting the experience or perceivable visual details)
- [ ] Created a ticket for the [help.shopify.com](https://help.shopify.com) documentation team about updates to theme settings. (Internal-only task)
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

sofiamatulis commented on snippets/loading-overlay.liquid:
```diff
@@ -1,15 +1,20 @@
-      <circle class="path" fill="none" stroke-width="6" cx="33" cy="33" r="30"></circle>
-    </svg>
-  </div>
+<div class="loading-overlay__spinner hidden">
```
thanks for clarifying! 👍 Looks good

sofiamatulis reviewed (commented):
Looks good 🎉 Great work

Just added a question about naming 👀

sofiamatulis commented on snippets/loading-overlay.liquid:
```diff
@@ -1,15 +1,20 @@
+{% comment %}
+  Renders loading-overlay spinner.
```
Yeah, exactly why I brought this up 🤔 no overlay.

@eugenekasimov let's rename it all :)

eugenekasimov replied:
Thanks team for great discussion, I renamed everything 👍

sofiamatulis commented on sections/main-collection-product-grid.liquid:
```diff
@@ -130,7 +119,6 @@
       >
         {%- if collection.products.size == 0 -%}
           <div class="collection collection--empty page-width" id="product-grid" data-id="{{ section.id }}">
-            <div class="loading-overlay gradient"></div>
```
It looks like it's related to this comment maybe? 🤔 but i cant open the screenshots

https://github.com/Shopify/dawn/issues/1558#issuecomment-1083992298

eugenekasimov replied:
Nice catch @andrewetchen! I put them back 👍

eugenekasimov commented on assets/component-facets.css:
```diff
@@ -931,20 +931,6 @@ input.mobile-facets__checkbox {
   visibility: hidden;
 }
 
-.product-count .loading-overlay__spinner,
-.product-count-vertical .loading-overlay__spinner {
-  display: none;
-  position: absolute;
-  right: 0;
-  top: 50%;
-  transform: translateY(-50%);
-  width: 1.8rem;
-}
-
-.product-count__text.loading + .loading-overlay__spinner {
-  display: block;
-}
-
```
@andrewetchen referring to you [comment](https://github.com/Shopify/dawn/pull/2996#pullrequestreview-1640715481) I removed it in my next [commit](https://github.com/Shopify/dawn/pull/2996/commits/62f2ff8fcef79503b2d34297fd9caba777e04502)

eugenekasimov commented:
> Recent changes look good so far. We still need to cover a few more cases then we should be good! 😄
> 
> 1. I'm still seeing a few instances of `loading-overlay__spinner` in the `component-facets.css` file.
> 
> These should be `loading__spinner`:
> 
> https://github.com/Shopify/dawn/blob/a53537bda8d9cdafbbb496ade271ab791bd96186/assets/component-facets.css#L934-L946
> 
This was completely removed [here](https://github.com/Shopify/dawn/pull/2996/commits/62f2ff8fcef79503b2d34297fd9caba777e04502). 
> 2. We should rename the `component-loading-overlay.css` file to `component-loading-spinner.css`.
> 
> And everywhere we have:
> 
> ```
> {{ 'component-loading-overlay.css' | asset_url
> ```
> 
> Change to:
> 
> ```
> {{ 'component-loading-spinner.css' | asset_url
> ```
I'm not sure about it because the file `component-loading-overlay.css` still has both loading-overlay and spinner styles. 
> 3. And when I search for any leftover instances of the string `loading-overlay`, I see the following two instances for `loading-overlay--error`. I think we should apply the same renaming here so it becomes `loading__spinner--error` 👍🏼
> 
> https://github.com/Shopify/dawn/blob/62f2ff8fcef79503b2d34297fd9caba777e04502/assets/quick-order-list.css#L334-L336
> 
> https://github.com/Shopify/dawn/blob/62f2ff8fcef79503b2d34297fd9caba777e04502/assets/quick-order-list.js#L249
> 

That makes sense to me. [Changed](https://github.com/Shopify/dawn/pull/2996/commits/dd90eea814d6d751caa4e3241e6871a3976bd6e5). 
> I think that captures everything from a renaming perspective ✨

eugenekasimov commented:
> @eugenekasimov Let's make sure to fix this bug as well :) related to loading overlay CSS
> 
> #3011 cc @andrewetchen

I think this bug is already fixed in this PR because now we use the snippet and the snippet has link to {{ 'component-loading-overlay.css' | asset_url | stylesheet_tag }}

sofiamatulis commented on sections/main-search.liquid:
```diff
@@ -77,66 +77,77 @@
+              {%- if settings.predictive_search_enabled -%}
+                <div class="predictive-search predictive-search--search-template" tabindex="-1" data-predictive-search>
+                  <div class="predictive-search__loading-state">
+                    <svg
```
Should we be using the snippet here?

sofiamatulis commented on sections/main-collection-product-grid.liquid:
```diff
@@ -1,5 +1,5 @@
 {{ 'template-collection.css' | asset_url | stylesheet_tag }}
-{{ 'component-loading-overlay.css' | asset_url | stylesheet_tag }}
+{{ 'component-loading-spinner.css' | asset_url | stylesheet_tag }}
```
Why are we importing this css here?

eugenekasimov commented on sections/main-collection-product-grid.liquid:
```diff
@@ -85,17 +85,6 @@
                 {%- endif -%}
               </span>
             </h2>
-            <div class="loading-overlay__spinner">
```
Okay, I talked to Meli and she talked to Dan Van, so they confirmed that that was a bug and we want show spinner during overlay here. I'm going to have a pairing with Meli on Monday just to make sure the spinner looks good. Meanwhile, I'll be working to fix this bug.

@sofiamatulis thanks for bringing this concern 👍

eugenekasimov commented:
I pushed a few updates and wanted to summarize what have been changed:

1. I fixed a bug when the spinner didn't show up for the facets on collections and search pages (details in the description to this PR).
2. I fixed a bug when the overlay covers not just product grid but also the search form and filters (details in the description to this PR).
3. I removed all instances when we were importing `component-loading-spinner.css` file to liquid files because now we import this file to the snippet.

eugenekasimov commented on snippets/cart-drawer.liquid:
```diff
@@ -218,20 +218,7 @@
-                            </div>
-                          </div>
-
+                          {%- render 'loading-spinner', cart_drawer: true -%}
```
forgot to remove 👍

eugenekasimov commented on sections/header.liquid:
```diff
@@ -5,7 +5,7 @@
 {%- if settings.predictive_search_enabled -%}
   <link rel="stylesheet" href="{{ 'component-price.css' | asset_url }}" media="print" onload="this.media='all'">
-  <link rel="stylesheet" href="{{ 'component-loading-overlay.css' | asset_url }}" media="print" onload="this.media='all'">
+  <link rel="stylesheet" href="{{ 'component-loading-spinner.css' | asset_url }}" media="print" onload="this.media='all'">
```
nope

sofiamatulis commented on assets/component-facets.css:
```diff
@@ -927,24 +927,21 @@ input.mobile-facets__checkbox {
 
-.product-count__text.loading {
+#ProductCountDesktop.loading,
+#ProductCount.loading {
```
thanks for clarifying!

sofiamatulis commented on assets/template-collection.css:
```diff
@@ -11,6 +11,25 @@
   }
 }
 
+.loading-overlay {
```
Going to break it down into 2 parts 👀 let me know your thoughts. Just trying to make sure this is in the right place and targeted with the right classes

1 - What I mean is that this `template-collection.css` is pulled into feat collection and other files. But here you are targetting all `loading-overlay`. Should it be something like `.template-collection-class .loading-overlay` ? If not this should live in the `loading-overlay` css file

2- `[template-collection.css` is pulled into `main-search`, `feat-collection` and `main-collection-prod-grid`. Is this needed in all of them?

eugenekasimov commented on sections/predictive-search.liquid:
```diff
@@ -1,149 +1,225 @@
-        <circle class="path" fill="none" stroke-width="6" cx="33" cy="33" r="30"></circle>
-      </svg>
-    </div>
+    {% render 'loading-spinner', predictive_search: true %}
```
It's not going to work this way because if we just wrap the snippet with <div class=''predictive-search__loading-state"></div> we will have

```
<div class="predictive-search__loading-state">
  <div class="loading__spinner hidden">
    <svg
      aria-hidden="true"
      focusable="false"
      class="spinner"
      viewBox="0 0 66 66"
      xmlns="http://www.w3.org/2000/svg"
    >
     <circle class="path" fill="none" stroke-width="6" cx="33" cy="33" r="30"></circle>
   </svg>
  </div>
</div>
```

There will be conflicts between `.predictive-search__loading-state` and `.loading__spinner` because they describe styles for the spinner and between `.hidden`

eugenekasimov replied:
As we agreed on a meeting I added a parameter **class** for the snippet so in the future if we need to add class(es) different from the default ones we can easily pass them via the parameter.

eugenekasimov commented:
> I was doing some visual testing with different settings and noticed some odd behaviour with the loading spinner not showing as expected on the collection page.
> 
> **Replication steps**
> 
> * Enable Horizontal or Vertical filters on the collection page > Product grid
> * Enable Cart type drawer
> * Navigate to a product and add it to the cart
> * Navigate back to the collection page
> * Select and deselect a filter (In stock for example) and notice the missing loading spinner
> 
> When cart type is set to either `Page` or `Popup notification`, it works as expected. I think there could be something potentially overriding when the cart type drawer is enabled: possibly something from `component-cart-drawer.css` or `component-cart.css` but can't confirm 🤔
> 
>  collection-page-loading-spinner-cart-drawer.mp4

Hey @andrewetchen , thanks for flagging this. The issue was I was targeting only first `.loading__spinner` while we might have a multiple places with spinner on the same page as an example the spinner in facets and cart-drawer. I made my targeting more specific for facets and covered all scenarios for different filter types.

sofiamatulis commented:
@eugenekasimov Could you resolve the conflicts before doing the final pass? :) https://screenshot.click/06-16-shclv-xrk32.png

eugenekasimov commented:
> @eugenekasimov Could you resolve the conflicts before doing the final pass? :) https://screenshot.click/06-16-shclv-xrk32.png

done!

eugenekasimov commented on assets/component-search.css:
```diff
@@ -68,3 +68,7 @@
   -webkit-appearance: none;
 }
+
+.template-search__results {
+  position: relative;
+}
```
Yes

eugenekasimov commented on sections/main-product.liquid:
```diff
@@ -220,19 +219,7 @@
-                        </svg>
-                      </span>
-                    </span>
+                    {%- render 'loading-spinner' -%}
```
Interesting but this is happening in main too
https://screenshot.click/06-02-mqulf-5oouq.mp4

In your case it's not happening in main because the price block was disabled.

sofiamatulis replied:
Good catch! I agree- since it's not related to this PR let's create a separate issue

sofiamatulis reviewed (approved):
Looks good 🎉 thanks for the changes and working on the feedback

eugenekasimov commented:
> 2. This also occurs on `main` so nothing introduced in this PR. The `<span>` for quantity rules on the PDP also gets a loading spinner when you adjust the cart item quantity within the cart drawer:

Let's not touch it here since this is not straight forward and not about refactoring.
